src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java

```diff
@@ -117,13 +117,26 @@ public long find(String name) {
     @SuppressWarnings("removal")
     public NativeLibrary loadLibrary(Class<?> fromClass, File file) {
         // Check to see if we're attempting to access a static library
+
         String name = findBuiltinLib(file.getName());
         boolean isBuiltin = (name != null);
         if (!isBuiltin) {
             name = AccessController.doPrivileged(new PrivilegedAction<>() {
                     public String run() {
                         try {
                             if (loadLibraryOnlyIfPresent && !file.exists()) {
+                                //Remove member name in brackets from file pathname, as such pathnames do not exist.
+                                //Original pathname with bracket is returned,which is handled by dlopen() in AIX.
+                                if (file.getName().contains("(")){
```

Comment:
Hello @suchismith1993, can you provide us some context on what this check is about? Is some code application code calling `System.load("foo(bar");` and this code in then trying to remove that `(`?

Adding this code here wouldn't be right. Plus, I see that in this new `if` block, the code then goes and renames the underlying `File` to a new name, which too isn't right.

Knowing the context of why this is needed would help us understand if anything needs to change here.

Reply:
In AIX, we have an usecase where shared libraries have certain member objects to be referred to. 
E.g libclang.a(shr_64.o) .  
When we provide the path to loadLibrary(), it searches if the file exists as absolute path .i.e libclang.a(shr_64.o), but such a path doesn't exist, only dlopen() would be able to recognise the member object mentioned in the path and able to load it. 

So before the code transitions to native code where dlopen() happens, the member object is being removed from the filename,so that the right path is referred(libclang.a) , and then the original path name with member name is passed on to the dlopen()

Reply:
> In AIX, we have an usecase where shared libraries have certain member objects to be referred to. E.g libclang.a(shr_64.o) .

Would you happen to know any official documentation which explains that AIX syntax? 

> When we provide the path to loadLibrary()

Do you mean some application code is calling the `System.loadLibrary()` method with such values?

Reply:
> > In AIX, we have an usecase where shared libraries have certain member objects to be referred to. E.g libclang.a(shr_64.o) .
> 
> Would you happen to know any official documentation which explains that AIX syntax?
> 

https://www.ibm.com/docs/en/aix/7.2?topic=l-ld-command
Text : 

**autoload: path/file(member) | Automatically load the archive member.
-- | --**
Autoload is a  flag used to load archive members. The format of the path is mentioned.

Reply:
> Do you mean some application code is calling the `System.loadLibrary()` method with such values?

Yes we are trying to install liblcang and also jextract and it fails with errors.

Exception in thread "main" java.lang.UnsatisfiedLinkError: no clang in java.library.path
 
        at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2439)
        at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:916)
        at java.base/java.lang.System.loadLibrary(System.java:2063)
        at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.RuntimeHelper.<clinit>(RuntimeHelper.java:65)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.RuntimeHelper.%3cclinit%3e(RuntimeHelper.java:65))
        at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.constants$17.<clinit>(constants$17.java:46)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.constants$17.%3cclinit%3e(constants$17.java:46))
        at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion$MH(Index_h.java:5690)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion$MH(Index_h.java:5690))
        at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion(Index_h.java:5698)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion(Index_h.java:5698))
        at [org.openjdk.jextract@22/org.openjdk.jextract.clang.LibClang.version(LibClang.java:93)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.LibClang.version(LibClang.java:93))
        at [org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.run(JextractTool.java:362)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.run(JextractTool.java:362))
        at [org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.main(JextractTool.java:174)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.main(JextractTool.java:174))


The actual member is libclang.a(libclang.so.16)

Reply:
Hello @suchismith1993, I haven't fully caught up on the necessity of this change. The current proposed change, renames files in the underlying filesystem in the code flow of `System.loadLibrary()` and I don't think is the right thing to do irrespective of whether we do it in AIX specific code or common code.

Could you update the JBS issue description to clearly explain what (application?) code is causing what issue? I know you mentioned about jextract here, but it's not clear whether this is some issue in jextract or something else. Having all those details in the JBS issue will be helpful to decide what changes should be pursued.

Reply:
As for this:

> Yes we are trying to install liblcang and also jextract and it fails with errors.

> Exception in thread "main" java.lang.UnsatisfiedLinkError: no clang in java.library.path
>
    at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2439)
    at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:916)
    at java.base/java.lang.System.loadLibrary(System.java:2063)
    at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.RuntimeHelper.<clinit>(RuntimeHelper.java:65)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.RuntimeHelper.%3cclinit%3e(RuntimeHelper.java:65))
    at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.constants$17.<clinit>(constants$17.java:46)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.constants$17.%3cclinit%3e(constants$17.java:46))
    at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion$MH(Index_h.java:5690)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion$MH(Index_h.java:5690))
    at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion(Index_h.java:5698)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion(Index_h.java:5698))
    at [org.openjdk.jextract@22/org.openjdk.jextract.clang.LibClang.version(LibClang.java:93)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.LibClang.version(LibClang.java:93))
    at [org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.run(JextractTool.java:362)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.run(JextractTool.java:362))
    at [org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.main(JextractTool.java:174)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.main(JextractTool.java:174))


I had a quick look at the latest jextract code and I don't see some of the classes that are reported in that stacktrace. Which exact version of jextract is this? If jextract is what motivated this change, then I would recommend checking on jextract-dev mailing list first https://mail.openjdk.org/mailman/listinfo/jextract-dev to understand if this is an issue in jextract itself.

Reply:
> Hello @suchismith1993, I haven't fully caught up on the necessity of this change. The current proposed change, renames files in the underlying filesystem in the code flow of `System.loadLibrary()` and I don't think is the right thing to do irrespective of whether we do it in AIX specific code or common code.
> 
> Could you update the JBS issue description to clearly explain what (application?) code is causing what issue? I know you mentioned about jextract here, but it's not clear whether this is some issue in jextract or something else. Having all those details in the JBS issue will be helpful to decide what changes should be pursued.

This is not specific to an application. We are looking to support member objects of AIX shared libraries in the Java ClassLoader. 
As for the renaming, the renaming is necessary because the member object is not part of the actual filename. 
Example : libclang.a(libclang.so.16) 
This member object, is identified only by the dlopen() in the hotspot code. But there is no actual file named libclang.a(libclang.so.16) in the filesystem. 
So when the check is done if file exists, it fails,i.e it checks for libclang.a(libclang.so.16).So i remove the braces and member name to extract the actual name of the file in the filesystem, which is libclang.a. But when welibclang.a and the dlopen ,it will fail to open it. 
Hence i need to rename the file to the name with braces before it passes to dlopen.So i reconstruct it to libclang.a(libclang.so.16)

Reply:
> As for this:
> 
> > Yes we are trying to install liblcang and also jextract and it fails with errors.
> 
> > Exception in thread "main" java.lang.UnsatisfiedLinkError: no clang in java.library.path
> 
> ```
> at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2439)
> at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:916)
> at java.base/java.lang.System.loadLibrary(System.java:2063)
> at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.RuntimeHelper.<clinit>(RuntimeHelper.java:65)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.RuntimeHelper.%3cclinit%3e(RuntimeHelper.java:65))
> at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.constants$17.<clinit>(constants$17.java:46)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.constants$17.%3cclinit%3e(constants$17.java:46))
> at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion$MH(Index_h.java:5690)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion$MH(Index_h.java:5690))
> at [org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion(Index_h.java:5698)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.libclang.Index_h.clang_getClangVersion(Index_h.java:5698))
> at [org.openjdk.jextract@22/org.openjdk.jextract.clang.LibClang.version(LibClang.java:93)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.clang.LibClang.version(LibClang.java:93))
> at [org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.run(JextractTool.java:362)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.run(JextractTool.java:362))
> at [org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.main(JextractTool.java:174)](mailto:org.openjdk.jextract@22/org.openjdk.jextract.JextractTool.main(JextractTool.java:174))
> ```
> 
> I had a quick look at the latest jextract code and I don't see some of the classes that are reported in that stacktrace. Which exact version of jextract is this? If jextract is what motivated this change, then I would recommend checking on jextract-dev mailing list first https://mail.openjdk.org/mailman/listinfo/jextract-dev to understand if this is an issue in jextract itself.

Jextract is one use case that prompted to make this change. There is no issue with the Jextract in particular. We probably might need a small change in the Index java file. But there are use cases where we need to support AIX dynamic archives to be able to load using Java ClassLoader,which is more like a front end to the hotspot code.
This support got recently added to hotspot code too. It is important to have it at the classloader level too.
@JoKern65  Could you add in more on the importance of this and provide your inputs ?

Reply:
> But there is no actual file named libclang.a(libclang.so.16) in the filesystem.
So when the check is done if file exists, it fails,i.e it checks for libclang.a(libclang.so.16).

I'll let Mandy and others more knowledgable of this area to correct me. From what I know, there was a change done a few releases back in the JDK (https://bugs.openjdk.org/browse/JDK-8275703) which accomodated macos platform where it too doesn't actually have the file on the filesystem in recent versions of macos. That change allowed platform specific code to dictate whether the file existence check needs to be done here in this loadLibrary code. That's done through the call to `jdk.internal.loader.ClassLoaderHelper.loadLibraryOnlyIfPresent()`. For macos, that implementation of that method resides in macos specific class here https://github.com/openjdk/jdk/blob/master/src/java.base/macosx/classes/jdk/internal/loader/ClassLoaderHelper.java#L41. For AIX too then, perhaps you could add an implementation which returns `false` (either always or in specific cases) and experiment with that to see if that's enough?

Reply:
> For AIX too then, perhaps you could add an implementation which returns false (either always or in specific cases) and experiment with that to see if that's enough?

I see that you actually have a ClassLoaderHelper in this PR for AIX, but that implementation in `loadLibraryOnlyIfPresent()` is currently returning `true` and thus enforcing the file existence check. Was it intentional to have it to `true`? Given the context of this change, shouldn't it be returning `false`?

Reply:
Correct. Had kept it as true, as it was in J9. but changing it to false works. 
So that would mean we are letting the java Runtime to handle paths that don't exist ?